packages/api-server/src/cache/data.ts

```diff
@@ -0,0 +1,206 @@
+    this.lock = args.lock;
+    this.executeCallResult = args.executeCallResult;
+    this.dataKeyExpiredTimeOut =
+      args.dataKeyExpiredTimeOutMs || DATA_KEY_EXPIRED_TIME_OUT_MS;
+
+    if (this.lock) {
+      this.lock.key.expiredTimeMs =
+        this.lock.key.expiredTimeMs || LOCK_KEY_EXPIRED_TIME_OUT_MS;
+      this.lock.pollIntervalMs = this.lock.pollIntervalMs || POLL_INTERVAL_MS;
+      this.lock.pollTimeOutMs = this.lock.pollTimeOutMs || POLL_TIME_OUT_MS;
+      this.lock.subscribe.timeOutMs =
+        this.lock.subscribe.timeOutMs || SUB_TIME_OUT_MS;
+    }
+  }
+
+  async get(dataKey: string) {
+    const value = await pubClient.get(dataKey);
+    if (value !== null) {
+      console.debug(
+        `[${this.constructor.name}]: hit cache via Redis.Get, key: ${dataKey}`
+      );
+      return value;
+    }
+
+    const setDataKeyOptions = { PX: this.dataKeyExpiredTimeOut };
+
+    if (this.lock == undefined) {
+      const result = await this.executeCallResult();
+      // set data cache
+      await pubClient.set(dataKey, result, setDataKeyOptions);
+      return result;
+    }
+
+    // use redis-lock for data cache
+    const t1 = new Date();
+    const lockValue = getLockUniqueValue();
+    const setLockKeyOptions = {
+      NX: true,
+      PX: this.lock.key.expiredTimeMs,
+    };
+
+    const releaseLock = async (lockValue: string) => {
+      if (!this.lock) throw new Error("enable lock first!");
+
+      const value = await pubClient.get(this.lock.key.name);
+      if (value === lockValue) {
+        // only lock owner can delete the lock
+        const delNumber = await pubClient.del(this.lock.key.name);
+        console.debug(
+          `[${this.constructor.name}]: delete key ${this.lock.key.name}, result: ${delNumber}`
+        );
+      }
+    };
+
+    while (true) {
+      const value = await pubClient.get(dataKey);
+      if (value !== null) {
+        console.debug(
+          `[${this.constructor.name}]: hit cache via Redis.Get, key: ${dataKey}`
+        );
+        return value;
+      }
+
+      const isLockAcquired = await pubClient.set(
+        this.lock.key.name,
+        lockValue,
+        setLockKeyOptions
+      );
+
+      if (isLockAcquired) {
+        try {
+          const result = await this.executeCallResult();
+          // set data cache
+          await pubClient.set(dataKey, result, setDataKeyOptions);
+          // publish the result to channel
+          const totalSubs = await pubClient.publish(
+            this.lock.subscribe.channel,
+            result
+          );
+          console.debug(
+            `[${this.constructor.name}]: publish message ${result} on channel ${this.lock.subscribe.channel}, total subscribers: ${totalSubs}`
+          );
+          await releaseLock(lockValue);
+          return result;
+        } catch (error) {
+          console.debug(error);
+          await releaseLock(lockValue);
+        }
+      }
+
+      // if lock is not acquired
+      try {
```

Comment:
should be moved to `else` block? otherwise, when an error is caught in `if (isLockAcquired)` block (for example, backend godwoken service is down), I think it will go to a dead loop.

Reply:
> should be moved to `else` block? otherwise, when an error is caught in `if (isLockAcquired)` block (for example, backend godwoken service is down), I think it will go to a dead loop.

I think in this case the [lock will be release](https://github.com/nervosnetwork/godwoken-web3/pull/156/files#diff-03bc3b7de02c3fd410085215655392a45a3696958610bde9668514dbda132289R143) so should be ok?